### DISCUSSION TITLE
No RGB visualizer ?

### DISCUSSION THREAD

**Discussion by Kimaul37** [aleab/acav-we#14] [Q&A] [answered]
Hello !
I was tinkering with your awesome visualizer to make me a beautiful moving wallpaper.
I wanted to make a wave visualizer that had a different RGB color depending on the power of the bass at this place on it and with a gradient to link the colors of all the points in the wave.
I can't seem to find how to do it after testing some things.
Am I just doing things wrong to be able to create that or is it not available at the moment ?
If it's the latter, would you consider adding a way to make such things with the parameters in the near future ?
Thanks in advance for your help !

**Comment by aleab** [comment, marked as answer]
Mmmhh... Let's see If I understand this correctly.

You want the following:
- every "point" of the wave shares the same color
- the color should change based on the power of the bass (or any other particular frequency range)
- the way it changes is a simple mapping from the power of the bass -> a color in a gradient

For example, if the power of the bass is a value from 0 to 100 and the gradient is Blue~Yellow, the color for a bass value of 0 should be exactly Blue, for a value of 100 should be exactly Yellow, and for anything in between it should be a different blending of the two colors; say the bass value is 30, then the resulting color would be "30% yellow" + "70% blue". (see [here](https://meyerweb.com/eric/tools/color-blend/#0000FF:FFFF00:10:hex) for example)

Is this correct?

In any case, there's currently no way of making the color of the visualizer react to the overall power of any particular frequency range; every "point" in the wave reacts purely based on its own value/altitude of the frequency "bucket" it represents.
I don't think it would be too hard to implement though.

**Reply by Kimaul37** [reply]
I'll answer with your words to be more understandable (sorry if the first message wasn't perfectly clear, I'm french)
Your second and third items of the list are pretty much correct but not the first.
I want every "point" of the wave to have its own color corresponding to the height of its frequency (same as how is determined the height of the wave at this same point). And to link these different points together on the wave, there should be a smooth color transition made between the ones that are next to each other.
The colors used could be either between selected colors and make a gradient with those like you said or we could choose to use a preset gradient you would have made that would look similar to the multicolor ones that can for example be used on RGB keyboards (and you could make some others if you have ideas). I would personally use this second option.

To summarize, the bass frequency determines the color displayed in a gradient that would be a preset or selected between 2 colors and each point of the wave (or any other linked type of visualizer we can do) would have its own color picked using that gradient with a formula using minimum bass value setting the minimum color value and same for maximum or any middle value between those. Also, as there are spaces between 2 points of the wave since it's not continuous, the color change between each one could be smoothed using the type of gradients you linked in your answer above for each transition. It would maybe take a certain amount of calculation but that should be OK for good PCs and less powerful ones could simply deactivate these smooth transitions.

I hope it's much clearer now that I have detailed it and used the correct words to describe.
If you know / can estimate, how many time do you think it could take for it to have a first iteration available ?

**Reply by aleab** [reply]
Before we continue, take a look at the following video clip.
Ignoring the fact that it's not a wave and that there's no smoothing between adjacent colors, is this the kind of effect you are looking for?

https://user-images.githubusercontent.com/11708634/214895709-1c505ce8-9165-4c0d-b391-0aafde666671.mp4

In the video the height of every bar (which in this context is the same as the intensity of the frequency it represents) is being mapped to a color in the [hue](https://en.wikipedia.org/wiki/Hue) range going from Red to Magenta.

**Reply by Kimaul37** [reply]
Yep, that's exactly that style of effect I'd like for the RGB (and this should not be displayed if it's completely silent but this option is already here in your wallpaper).
Also, maybe we should be able to use this preset and then, if we want, select a less wide range in it so you could do a double slider (or an other way if that's not possible with WE UI) to choose our limits in the spectrum. Someone could not want to have the red for example but the rest would be fine.
On a completely different subject, I tried to update my Spotify token with your link provided earlier today but when clicking the Request button on the page, it returned a 404 error marked in red below the token bar so I think you have something to fix there.

**Reply by aleab** [reply]
Awesome! Then you'll be happy to know that it was already possible to do this: [example](https://i.imgur.com/VFuDcRA.png).

The "Color" setting changes the start of the range, and "Max Color" changes the end. It only affects the hue, so the brightness and saturation of any intermediate color will be the same as that of the starting color.
Also, the algorithm does not loop over the hue spectrum range, meaning that if you select Magenta as starting color and Yellow as max, the algorithm will not generate any red color (it won't go to the right and loop over, but will instead go from right to left).

![Hue range](https://user-images.githubusercontent.com/11708634/214970901-9fdaada4-c202-4507-971f-6dbf90004537.png)

The only thing left that is not currently implemented is color smoothing. The curve on the left is how it looks now, the one on the right is how you would want it. Correct?

<img src="https://user-images.githubusercontent.com/11708634/214975730-cb794768-0b59-495b-9fab-d01753700ee5.png" height="150"/>

> On a completely different subject, I tried to update my Spotify token with your link provided earlier today but when clicking the Request button on the page, it returned a 404 error marked in red below the token bar so I think you have something to fix there.

Fixed. Thank you!

**Comment by Kimaul37** [comment]
OK ! That's cool news !
I just set that up with the parameters on your image and orange to pink selection and it kinda works
Yes, totally correct.
It's a bit weird without smoothing on the wave for the moment xD
![image](https://user-images.githubusercontent.com/28737149/215119335-102040df-841d-4a3f-8d37-8ef865a8174b.png)
(The visualizer at the center of the image seems small because I have two 2K displays and cropped the image from a full width screenshot xD)

Thank you for your useful answers and I hope to quickly see the smoothing added (and also maybe presets of colors for things to be simpler)
I give you my best wishes for the future of your projects (but this one above all xD) and yours as a person
And also thanks for the token thing fix.

**Reply by Kimaul37** [reply]
Here's a V2 for my wallpaper ^^
![image](https://user-images.githubusercontent.com/28737149/215123199-69ce6c05-0565-48fb-a449-a1ca7a797502.png)

**Reply by aleab** [reply]
That's a very nice wallpaper!
I'll start looking into the color smoothing over the weekend.

Also, thank you for the kind words.

**Comment by aleab** [comment]
Done! I just published the new version to Steam.
No presets yet. I'm still thinking on how I want to do that.

https://user-images.githubusercontent.com/11708634/215841915-1a4ecbef-a28f-4808-9e16-ec4ae79a9d72.mp4

**Reply by Kimaul37** [reply]
Thanks ! Your wallpaper is now perfect for my use case !
No problem for the time needed on the presets, it was a suggestion I made much more for the other users than for me, mine is already pretty much ready
Keep going with the good work and I wish you my best wishes for the future of it !

**Comment by darkshadow-666** [comment]
hello i really like the design i just whant to know where can i get it thanks
